lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+/**
+ * A custom REST server for Gutenberg.
+ *
+ * @package gutenberg
+ * @since   6.8.0
+ */
+
+class Gutenberg_Hierarchical_Sort {
+
+
+	private static $post_ids = array();
+	private static $levels   = array();
+	private static $instance;
+
+	public static function init() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	public function run( $args ) {
+		$new_args = array_merge(
+			$args,
+			array(
+				'fields'         => 'id=>parent',
+				'posts_per_page' => -1,
+			)
+		);
+		$query    = new WP_Query( $new_args );
+		$posts    = $query->posts;
+		$result   = self::sort( $posts );
+
+		self::$post_ids = $result['post_ids'];
+		self::$levels   = $result['levels'];
```

Comment:
Why are we using static variables instead of returning the results instead? Personally I don't like this kind of caching. Actually, it's really not clear to me why this whole class exists. It can be just a function no (or a couple)?

Reply:
The only reason for this class to exist is to make this PR performant. Given this PR uses two different filters, we'd have to calculate the info twice if we don't have a way to hold the results in the interim. The implementation of the backport PR would be a bit different that this: it can just modify the [get_items](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L215) method directly, so this can be simplified to be a simple function.

Reply:
More context: I used the filter approach in this PR instead of modifying the `get_items` method directly because I wanted reviewers to understand what changed. If this modified the `get_items` method it'd have been hundreds of lines of (untouched) code with a few tweaks in between.

Reply:
I appreciate the review-friendly approach, especially when approaching something new to DataViews like hierarchy. The current approach with the singleton class isn't perfect, but I think it's acceptable for now. Thinking about Riad's question, what crossed my mind was to use global functions with an explicit cache:

```php
// In rest_page_query filter
$cache_key        = serialize( $request->get_params() );
$result           = get_hierarchical_sort( $args, $cache_key );
$args['post__in'] = $result['post_ids'];

// In rest_prepare_page filter
//
// The `null` argument combines with the cache key to signal that the function
// should retrieve the previously calculated value.
$cache_key               = serialize( $request->get_params() );
$result                  = get_hierarchical_sort( null, $cache_key );
$response->data['level'] = $result['levels'];
```

But, honestly, this comes with other problems (a false impression that nothing could have mutated in between the two function calls?) and I don't think we should invest more time in anything other than "the real thing", as André points out, i.e. `get_items`. Until then, this class is fine. Thoughts?

Reply:
I still don't like the class to be honest. The same reason I don't like the current theme_json class. We know by experience now that this architecture created a lot of bugs (outdates styles...). I don't think using classes properties to cache things like that is good.

You call a run function with `args` but the cache is actually independent of the "args", so one calling the run function later with another set of args will conflict with the initial call.

Reply:
What kind of performance penalty are we talking about if we remove all caching?

If indeed there is a penalty to avoid, how palatable would something like this be to both of you? (This works in my local Pages view.)

```diff
diff --git a/lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php b/lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php
index f61002f435a..7d9d9b9c22e 100644
--- a/lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php
+++ b/lib/compat/wordpress-6.8/class-gutenberg-hierarchical-sort.php
@@ -8,19 +8,7 @@
  */
 
 class Gutenberg_Hierarchical_Sort {
-	private static $post_ids = array();
-	private static $levels   = array();
-	private static $instance;
-
-	public static function get_instance() {
-		if ( null === self::$instance ) {
-			self::$instance = new self();
-		}
-
-		return self::$instance;
-	}
-
-	public function run( $args ) {
+	public static function run( $args ) {
 		$new_args = array_merge(
 			$args,
 			array(
@@ -32,8 +20,7 @@ class Gutenberg_Hierarchical_Sort {
 		$posts    = $query->posts;
 		$result   = self::sort( $posts );
 
-		self::$post_ids = $result['post_ids'];
-		self::$levels   = $result['levels'];
+		return $result;
 	}
 
 	/**
@@ -146,14 +133,6 @@ class Gutenberg_Hierarchical_Sort {
 			}
 		}
 	}
-
-	public static function get_post_ids() {
-		return self::$post_ids;
-	}
-
-	public static function get_levels() {
-		return self::$levels;
-	}
 }
 
 add_filter(
@@ -171,17 +150,23 @@ add_filter(
 add_filter(
 	'rest_page_query',
 	function ( $args, $request ) {
+		global $_gutenberg_hierarchical_sort_cache;
+
 		if ( ! Gutenberg_Hierarchical_Sort::is_eligible( $request ) ) {
 			return $args;
 		}
 
-		$hs = Gutenberg_Hierarchical_Sort::get_instance();
-		$hs->run( $args );
+		$result = Gutenberg_Hierarchical_Sort::run( $args );
 
 		// Reconfigure the args to display only the ids in the list.
-		$args['post__in'] = $hs->get_post_ids();
+		$args['post__in'] = $result['post_ids'];
 		$args['orderby']  = 'post__in';
 
+		$_gutenberg_hierarchical_sort_cache[] = array(
+			'request' => $request,
+			'result'  => $result,
+		);
+
 		return $args;
 	},
 	10,
@@ -195,8 +180,13 @@ add_filter(
 			return $response;
 		}
 
-		$hs                      = Gutenberg_Hierarchical_Sort::get_instance();
-		$response->data['level'] = $hs->get_levels()[ $post->ID ];
+		global $_gutenberg_hierarchical_sort_cache;
+		foreach ( $_gutenberg_hierarchical_sort_cache as $tuple ) {
+			if ( $request === $tuple['request'] ) {
+				$levels                  = $tuple['result']['levels'];
+				$response->data['level'] = $levels[ $post->ID ];
+			}
+		}
 
 		return $response;
 	},
```

Reply:
@youknowriad @mcsf I'd like to land the backport PR and it doesn't use filters or the static cache in the `HierarchicalSort` class. https://github.com/WordPress/wordpress-develop/pull/8014

Note the `HierarchicalSort` class still exist and exposes two public static methods. I think this is nice to prevent polluting the namespace with intermediate helper functions. If that's a blocker for you, I'm happy to make everything functions.

Reply:
> Note the HierarchicalSort class still exist and exposes two public static methods. I think this is nice to prevent polluting the namespace with intermediate helper functions. If that's a blocker for you, I'm happy to make everything functions.

Using a class is not a blocker, but I would love for us to have a system in place for when to use functions and when to use classes rather than each one coming up with its own approach to things and second guessing ourselves.